### PR TITLE
FEDOT Command Line Interface (CLI) prototype 

### PR DESCRIPTION
**Pull request by ChrisLisbon**
FEDOT CLI prototype supports all API parameters, which can be input from the command line.
It provides by running fedot_cli.py with flags described in help. 
Integration tests with imitation of console run was added.

### CODE REVIEW

**Review by Dreamlone — COMMENTED** [review]
А ты специально файлы из cases/industrial удалила? Они вроде для того нужны, чтобы при случае можно было продемонстрировать "нашим индустриальным партнерам" как FEDOT работает на примере вот этих сериализованных моделей

Но в целом, там примеры были для ipnb ноутбуков, а ноутбуки устарели, так что, думаю, в челом industrial надо бы обновить или убрать если никому не нужно
@nicl-nno, присоединяйся к обсуждению

File: cli_application/fedot_cli.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
+import argparse
```
Comment (Dreamlone) [inline cli_application/fedot_cli.py:1]: Может cli_application спрятать в fedot, или в fedot/api?

File: cli_application/fedot_cli.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+from fedot.core.repository.tasks import TsForecastingParams
+from fedot.api.main import Fedot
+import os
+
+
+def add_args_to_parser(parser, tag, help, required=False, is_list=False):
+    """ Function for formalized parser arguments creation
+    :param parser: ArgumentParser class object
+    :param tag: str, name of parameter to parse
+    :param help: str, help text for parameter
```
Comment (Dreamlone) [inline cli_application/fedot_cli.py:12]: Type hints лучше выше вынести а вот из doctring можно убрать

File: cli_application/fedot_cli.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    :param required: bool, is the parameter required or not
+    :param is_list: bool, does the parameter contain multiple input
+    """
+    if is_list is True:
+        nargs = '*'
+    elif is_list is False:
+        nargs = None
```
Comment (Dreamlone) [inline cli_application/fedot_cli.py:19]: Можно капельку проще
```
if is_list:
    nargs = '*'
else:
    nargs = None
```

**Comment by ChrisLisbon** [comment]
> А ты специально файлы из cases/industrial удалила? Они вроде для того нужны, чтобы при случае можно было продемонстрировать "нашим индустриальным партнерам" как FEDOT работает на примере вот этих сериализованных моделей
> 
> Но в целом, там примеры были для ipnb ноутбуков, а ноутбуки устарели, так что, думаю, в челом industrial надо бы обновить или убрать если никому не нужно @nicl-nno, присоединяйся к обсуждению

Я убрала потому что подумала, что это рудимент, тк папка была названа как временные файлы обычно называются у нас. Верну если это ваажная штука.

File: fedot/api/main.py [hunk 4]
```diff
@@ -41,12 +41,13 @@ class Fedot:
...
             'timeout' - composing time (minutes)
             'available_operations' - list of model names to use
-            'with_tuning' - allow huperparameters tuning for the model
+            'with_tuning' - allow hyperparameters tuning for the model
             'cv_folds' - number of folds for cross-validation
             'validation_blocks' - number of validation blocks for time series forecasting
             'initial_pipeline' - initial assumption for composing
             'genetic_scheme' - name of the genetic scheme
             'history_folder' - name of the folder for composing history
+            'metric' - metric for quality calculation during composing
```
Comment (Dreamlone) [inline fedot/api/main.py:50]: Решила все таки не трогать composing_params? Просто они вроде больше api_params, или fedot_params, а не composing. Да и в целом, думаю, можно в рамках этого PR сделать передачу аргументов немного попроще в API - через именованные аргументы 

Плюс еще вопрос: а metric именно для композера здесь, или в том числе и для тюнера?

**Review by Dreamlone — CHANGES_REQUESTED** [review]

File: fedot/api/main.py (hunk 4 above)
Comment (ChrisLisbon) [reply fedot/api/main.py:50]: В данном случае именно для композера.
api_params - звучат вроде норм, могу поменять.

эта метрика может быть в виде списка и для нее возвращаются сущности для таска, композера и тюнера
`metric_name = self.api_params['metric_name']`
`self.task_metrics, self.composer_metrics, self.tuner_metrics = self.metrics.get_metrics_for_task(metric_name)`
`self.api_params['tuner_metric'] = self.tuner_metrics`

Comment (Dreamlone) [reply fedot/api/main.py:50]: Окс

File: cli_application/fedot_cli.py (hunk 1 above)
Comment (ChrisLisbon) [reply cli_application/fedot_cli.py]: Перенесла в fedot/api

File: cli_application/fedot_cli.py (hunk 2 above)
Comment (ChrisLisbon) [reply cli_application/fedot_cli.py]: Перенесла

File: cli_application/fedot_cli.py (hunk 3 above)
Comment (ChrisLisbon) [reply cli_application/fedot_cli.py]: Поправила

File: fedot/api/fedot_cli.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+remove_none_keys(fit_params)
+
+main_params['composer_params'] = composer_params
+
+model = Fedot(**main_params)
+print("\nFitting start...")
+model.fit(**fit_params)
+print("\nPrediction start...")
+prediction = model.predict(features=getattr(parameters, 'test'), save_predictions=True)
+print(f"\nPrediction saved at {os.getcwd()}\\predictions.csv")
```
Comment (Dreamlone) [inline fedot/api/fedot_cli.py:156]: А это тут пример использования уже идёт? Если так, то его стоит отдельно держать от скрипта, чтобы он запускался при импорте чего-нибудь из этого файла. Ну и пример тогда стоит добавить в examples

Comment (ChrisLisbon) [reply fedot/api/fedot_cli.py:156]: Нет, это не пример, это собственно сам запуск федота из выбранных из консоли параметров. Пример есть в интеграционниках.

File: test/integration/cli_application/cli_call_test.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+def test_cli_classification():
+    """ Test executing classification problem from cli with saving prediction"""
+    bat_name = 'cli_classification_call.bat'
+    df = run_console(bat_name)
+    assert len(df) == 15
+
+
+def test_cli_regression():
+    """ Test executing regression problem from cli with saving prediction"""
```
Comment (Dreamlone) [inline test/integration/cli_application/cli_call_test.py:54]: А запуск в unit тестах даже самого коротенького примера с консолью - это слишком дорогое удовольствие? Просто можно было бы просто запустить с параметром predefined_model = 'auto' и api обучил бы всего один пайплайн

Comment (ChrisLisbon) [reply test/integration/cli_application/cli_call_test.py:54]: Дело не в дороговизне а в нереализуемости. Тк мы проверяем запуск из консоли, то идет эмуляция этой консоли с созданием отдельного процесса. Git такое не переварит ни при каком варианте.

File: fedot/api/fedot_cli.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        nargs = '*'
+    else:
+        nargs = None
+    parser.add_argument(tag, help=help, required=required, nargs=nargs)
+
+
+def remove_none_keys(parameters: dict):
+    """ Function that removes nan parameters from input
+    and converts numbers from strings
+    :param parameters: argparse filtered keys
```
Comment (nicl-nno) [inline fedot/api/fedot_cli.py:26]: Если функция ещё и "converts numbers from strings" - то лучше назвать её preprocess_keys или как-то так.

File: test/integration/cli_application/cli_classification_call.bat [hunk 8]
```diff
@@ -0,0 +1,3 @@
+set python_path = "DEFAULT"
```
Comment (nicl-nno) [inline test/integration/cli_application/cli_classification_call.bat:1]: Может это в expamples? 
Вроде тут никакого тестирования результата нет.

File: test/integration/cli_application/cli_call_test.py (hunk 6 above)
Comment (nicl-nno) [reply test/integration/cli_application/cli_call_test.py:54]: Можно обернуть код fedot_cli.py в функцию и протестировать её, подменив консольный ввод, который ты собираешь в parameters.

File: test/integration/cli_application/cli_classification_call.bat (hunk 8 above)
Comment (ChrisLisbon) [reply test/integration/cli_application/cli_classification_call.bat:1]: Ну тут тестирование, что консоль не падает. А example написать довольно сложно потому что сама идея что это просто строка в консоли и задавать ее через файлы сомнительно

File: fedot/api/fedot_cli.py (hunk 7 above)
Comment (ChrisLisbon) [reply fedot/api/fedot_cli.py]: Поправила

**Comment by nicl-nno** [comment]
Хорошо бы ещё сделать страничку в документацию с примерами использования. По аналогии с https://fedot.readthedocs.io/en/latest/fedot/features/pipeline_json.html

File: test/integration/cli_application/cli_classification_call.bat (hunk 8 above)
Comment (nicl-nno) [reply test/integration/cli_application/cli_classification_call.bat:1]: В examples могут те  же самые bat-скрипты лежать.

File: cli_application/fedot_cli.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+parser.add_argument('--c_popsize',
+                    help='Composer parameter: population size')
+parser.add_argument('--c_gen_num',
+                    help='Composer parameter: number of generations')
+parser.add_argument('--c_timeout',
+                    help='Composer parameter: composing time (minutes)')
+parser.add_argument('--c_opers',
+                    help='Composer parameter: model names to use',
+                    nargs='*')
+parser.add_argument('--c_tuning',
```
Comment (valer1435) [inline cli_application/fedot_cli.py]: Может быть сделать это просто необязательным аргументом? 
Кажется проще писать --with_tuning, чем всегда --c_tuning <num>

File: test/integration/cli_application/cli_call_test.py (hunk 6 above)
Comment (ChrisLisbon) [reply test/integration/cli_application/cli_call_test.py]: Переписала fedot_cli в виде функций и добавила с ними юнит тест (test_api_cli_params.py)

File: test/integration/cli_application/cli_classification_call.bat (hunk 8 above)
Comment (ChrisLisbon) [reply test/integration/cli_application/cli_classification_call.bat]: Перенесла в папку examples/cli_application

File: cli_application/fedot_cli.py (hunk 9 above)
Comment (ChrisLisbon) [reply cli_application/fedot_cli.py]: Он и так необязательный, если параметр не передается (или передается что либо кроме 1), то по умолчанию становится False

**Review by Dreamlone — APPROVED** [review]
Всё кул, только я бы все таки попросил бы тест немного упростить тест `test_cli_with_parameters`. Я понимаю, что написанные unit тесты хороши потому что позволяют разобраться как запускать FEDOT через консоль, но они работают очень долго - полминуты только на один тест на моем ноуте - долговато. Оставь пример для временного ряда и для классификации. Регрессия во многом дублирует задачу классификации по внешним аргументам (только problem classification нужно поменять на problem regression) 
У меня в таком случае тест отработал за 17 секунд

**Review by nicl-nno — APPROVED** [review]
Добавить страничку документации и норм.
